Sources/StreamChatUI/VoiceRecording/VoiceRecordingVC.swift

```diff
@@ -0,0 +1,739 @@
+                        waveform: self.content.waveform
+                    )
+                    self.stopRecordingButton.isHidden = true
+
+                    if self.sendImmediately {
+                        if self.components.isVoiceRecordingConfirmationRequiredEnabled {
+                            self.updateContentByApplyingAction(.confirm)
+                        } else {
+                            self.updateContentByApplyingAction(.send)
+                        }
+                        self.sendImmediately = false
+                        self.content = .idle
+                    }
+                case let .failure(error):
+                    log.error(error)
+                }
+            }
+        }
+    }
+
+    open func audioRecorder(
+        _ audioRecorder: AudioRecording,
+        didFailWithError error: Error
+    ) {
+        log.error(error)
+        content = .idle
+    }
+
+    // MARK: - AudioPlayingDelegate
+
+    open func audioPlayer(
+        _ audioPlayer: AudioPlaying,
+        didUpdateContext context: AudioPlaybackContext
+    ) {
+        let isActive = context.assetLocation == content.location
+
+        switch (isActive, context.state) {
+        case (true, .playing), (true, .paused):
+            liveRecordingView.content = .init(
+                isRecording: false,
+                isPlaying: context.state == .playing,
+                duration: context.duration,
+                currentTime: context.currentTime,
+                waveform: liveRecordingView.content.waveform
+            )
+        case (true, .stopped):
+            liveRecordingView.content = .init(
+                isRecording: false,
+                isPlaying: false,
+                duration: context.duration,
+                currentTime: 0,
+                waveform: liveRecordingView.content.waveform
+            )
+        case (false, _):
+            liveRecordingView.content = .init(
+                isRecording: false,
+                isPlaying: false,
+                duration: content.duration,
+                currentTime: 0,
+                waveform: liveRecordingView.content.waveform
+            )
+        default:
+            break
+        }
+    }
+
+    // MARK: - Private Action Handlers
+
+    @objc private func didTapDiscard(_ sender: UIButton) {
+        updateContentByApplyingAction(.discard)
+    }
+
+    @objc private func didTapStop(_ sender: UIButton) {
+        updateContentByApplyingAction(.stop)
+    }
+
+    @objc private func didTapPlayPause(_ sender: UIButton) {
+        if sender.isSelected {
+            updateContentByApplyingAction(.pause)
+        } else {
+            updateContentByApplyingAction(.play)
+        }
+    }
+
+    @objc private func didTapSend(_ sender: UIButton) {
+        updateContentByApplyingAction(.send)
+    }
+
+    @objc private func didTapConfirm(_ sender: UIButton) {
+        updateContentByApplyingAction(.confirm)
+    }
```

Comment:
Any reason why you are passing the sender as a parameter here?

Reply:
The `sender` provides context about the caller. It's required for the `didTapPlayPause` to know which action to make. As UIKit is going to pass it for free, when we addTarget and to be consistent across the methods, I included it to all of them. I'm happy to remove them if you think that they can cause issues.